File: sdks/csharp/src/NovaBilling/Customers/Requests/CreateCustomerDto.cs
using System.Text.Json.Serialization;
using NovaBilling.Core;

namespace NovaBilling;

[Serializable]
public record CreateCustomerDto
{
    /// <summary>
    /// Tenant's user ID
    /// </summary>
    [JsonPropertyName("externalId")]
    public required string ExternalId { get; set; }

    [JsonPropertyName("email")]
    public required string Email { get; set; }

    [JsonPropertyName("name")]
    public string? Name { get; set; }

    [JsonPropertyName("country")]
    public string? Country { get; set; }

    /// <summary>
    /// ISO currency code
    /// </summary>
    [JsonPropertyName("currency")]
    public required string Currency { get; set; }

    /// <summary>
    /// Custom metadata
    /// </summary>
    [JsonPropertyName("metadata")]
    public Dictionary<string, object?>? Metadata { get; set; }

    /// <summary>
    /// Net payment terms in days (overrides org and plan defaults)
    /// </summary>
    [JsonPropertyName("netPaymentTerms")]
    public double? NetPaymentTerms { get; set; }

    /// <summary>
    /// Backdate createdAt (ISO 8601). For data imports.
    /// </summary>
    [JsonPropertyName("createdAt")]
    public string? CreatedAt { get; set; }

    /// <inheritdoc />
    public override string ToString()
    {
        return JsonUtils.Serialize(this);
    }
}

File: sdks/php/src/Customers/Requests/CreateCustomerDto.php
<?php

namespace NovaBilling\Customers\Requests;

use NovaBilling\Core\Json\JsonSerializableType;
use NovaBilling\Core\Json\JsonProperty;
use NovaBilling\Core\Types\ArrayType;

class CreateCustomerDto extends JsonSerializableType
{
    /**
     * @var string $externalId Tenant's user ID
     */
    #[JsonProperty('externalId')]
    public string $externalId;

    /**
     * @var string $email
     */
    #[JsonProperty('email')]
    public string $email;

    /**
     * @var ?string $name
     */
    #[JsonProperty('name')]
    public ?string $name;

    /**
     * @var ?string $country
     */
    #[JsonProperty('country')]
    public ?string $country;

    /**
     * @var string $currency ISO currency code
     */
    #[JsonProperty('currency')]
    public string $currency;

    /**
     * @var ?array<string, mixed> $metadata Custom metadata
     */
    #[JsonProperty('metadata'), ArrayType(['string' => 'mixed'])]
    public ?array $metadata;

    /**
     * @var ?float $netPaymentTerms Net payment terms in days (overrides org and plan defaults)
     */
    #[JsonProperty('netPaymentTerms')]
    public ?float $netPaymentTerms;

    /**
     * @var ?string $createdAt Backdate createdAt (ISO 8601). For data imports.
     */
    #[JsonProperty('createdAt')]
    public ?string $createdAt;

    /**
     * @param array{
     *   externalId: string,
     *   email: string,
     *   currency: string,
     *   name?: ?string,
     *   country?: ?string,
     *   metadata?: ?array<string, mixed>,
     *   netPaymentTerms?: ?float,
     *   createdAt?: ?string,
     * } $values
     */
    public function __construct(
        array $values,
    ) {
        $this->externalId = $values['externalId'];
        $this->email = $values['email'];
        $this->name = $values['name'] ?? null;
        $this->country = $values['country'] ?? null;
        $this->currency = $values['currency'];
        $this->metadata = $values['metadata'] ?? null;
        $this->netPaymentTerms = $values['netPaymentTerms'] ?? null;
        $this->createdAt = $values['createdAt'] ?? null;
    }
}
